### ISSUE TITLE
parallel prange race condition

### ISSUE THREAD

**Issue by synapticarbors** [numba/numba#3681]
Having read the documentation on [Explicit Parallel Loops](https://numba.pydata.org/numba-doc/dev/user/parallel.html#explicit-parallel-loops) using `prange`, it was not immediately clear to me if using a `+=` reduction to alter the same element of a numpy array in different threads is supported. I'm using Numba v0.42.0 (tested using Python 2.7 on OSX and Python 3.7 on Windows) and the following minimal example demonstrates that when `parallel=True`, the function produces different results than the pure python or non-parallel numba function:

```python
from __future__ import print_function

import numba as nb
import numpy as np


def func2(x):
    n = x.shape[0]
    y = np.zeros_like(x)
    for i in nb.prange(n):
        for j in range(n):
            if i == j:
                continue

            y[0, 1] += x[i, j]

    return y


func2_nb = nb.jit(nopython=True, nogil=True, parallel=False)(func2)
func2_nb_p = nb.jit(nopython=True, nogil=True, parallel=True)(func2)


if __name__ == '__main__':
    for k in range(1000):
        n = 20
        X = np.random.normal(size=(n, n))

        # Pure python
        Y1 = func2(X)

        # numba non-parallel
        Y2 = func2_nb(X)

        # numba parallel
        Y3 = func2_nb_p(X)

        if not np.allclose(Y1, Y3):
            print(k, Y1.sum(), Y2.sum(), Y3.sum())
            break
```

Is this a bug or is this type of function not supported in parallel mode? The issue can be quite subtle and only appears intermittently if a specific element is being written to only a few times by each parallel thread, whereas if every loop writes to the same element, it happens consistently.

**Comment by stuartarchibald** [comment]
Thanks for the question. I do not think that this is supported and I seem to have a vague recollection of looking at this before, so went searching and found https://github.com/numba/numba/issues/3522#issuecomment-441302072 which is seemingly related.

The [documentation](https://numba.pydata.org/numba-doc/dev/user/parallel.html#explicit-parallel-loops) alludes to this but perhaps is not sufficiently clear/specific:
```
 Reductions in this manner are supported for scalars and for arrays of arbitrary dimensions.
```
suggestions to improve the clarity of this statement are most welcome.

**Comment by jonmmease** [comment]
Would it be accurate to say that this issue describes the problem that is addressed in OpenMP by [array reductions](http://dvalts.io/optimisation/code/2016/11/06/OpenMP-array_reduction.html)?

If so, is this a feature that would be within scope for numba as well?

**Comment by stuartarchibald** [comment]
@jonmmease IIRC the issue being described here is to do with parallel reductions into array slices, however OpenMP style array reductions are a similar problem. I think fixing both to work would be a similar code path.

**Comment by Yoshi2112** [comment]
Hi all,

I've just come across this thread after trying to work out why the particle code I'm trying to write sometimes doesn't give the correct results. Is there any progress on numba's ability to perform a parallel reduction with an array as a target? E.g. having millions of particles spread across a domain and summing these in parallel to a mesh of gridpoints? I know it's possible to do by hand by dividing the domain into the number of threads, etc., but as someone who isn't a programmer or computer scientist, that would take me a lot longer.

Thanks!

**Comment by stuartarchibald** [comment]
Closing as fixed in #4170. Thanks.
